### PR TITLE
IP-Adapter Safetensor Support

### PR DESCRIPTION
**Pull request by blessedcoolant**
## Summary

This PR adds support for IP Adapter safetensor files for direct usage inside InvokeAI.

# TEST

You can download the [Composition Adapters](https://huggingface.co/ostris/ip-composition-adapter) which weren't previously supported in Invoke and try them out. Every other IP Adapter model should work too.

If you pick a Safetensor IP Adapter model, you will also need to set ViT-H or ViT-G next to it. This is a raw implementation. Can refine it further based on feedback.

Prompt: `Spiderman holding a bunny` -- Exact same composition as the adapter image.
opera_UHlo1IyXPT

### CODE REVIEW

**Comment by blessedcoolant** [comment]
@lstein I would appreciate it if you could take at my checkpoint implementation for IP Adapters and let me know if all is in order. Stuff works as expected but I have not done this before and I just want to make sure I didn't miss anything.

**Review by hipsterusername — APPROVED** [review]

File: invokeai/backend/model_manager/load/model_loaders/ip_adapter.py [hunk 1]
```diff
@@ -31,7 +32,7 @@ def _load_model(
         if submodel_type is not None:
             raise ValueError("There are no submodels in an IP-Adapter model.")
         model = build_ip_adapter(
-            ip_adapter_ckpt_path=str(model_path / "ip_adapter.bin"),
+            ip_adapter_ckpt_path=str(model_path),
```
Comment (lstein) [inline invokeai/backend/model_manager/load/model_loaders/ip_adapter.py:35]: Foo. The model record isn't passed to `_load_model()`. I'll fix that. Here's a quick workaround.
```suggestion
            ip_adapter_ckpt_path=str(model_path if model_path.is_file() else model_path / "ip_adapter_bin"),
```

[the remark this answers is not in the document]
Comment (blessedcoolant) [reply invokeai/backend/model_manager/load/model_loaders/ip_adapter.py:35]: This was a carry over from the existing code. I removed the hardcoded file name here and pushed it inside the build function so I could account for both. I left the existing code untouched but like you said, it did seem off to me. I'll give it a test and see if I need to update this.

Comment (blessedcoolant) [reply invokeai/backend/model_manager/load/model_loaders/ip_adapter.py:35]: This is handled further down in the `load_ip_adapter_tensors` function of `backend/ip_adapter/ip_adapter.py` .. Unless I am understanding this incorrectly.

File: invokeai/app/invocations/ip_adapter.py [hunk 2]
```diff
@@ -48,20 +49,25 @@ class IPAdapterOutput(BaseInvocationOutput):
...
     ip_adapter_model: ModelIdentifierField = InputField(
         description="The IP-Adapter model.",
         title="IP-Adapter Model",
         input=Input.Direct,
         ui_order=-1,
         ui_type=UIType.IPAdapterModel,
     )
-
+    clip_vision_model: Literal["ViT-H", "ViT-G"] = InputField(
+        description="CLIP Vision model to use", default="ViT-H", ui_order=2
```
Comment (RyanJDick) [inline invokeai/app/invocations/ip_adapter.py:69]: The description should probably explain that the `clip_vision_model` will be overridden if the right choice is known for the selected `ip_adapter_model`

File: invokeai/backend/ip_adapter/ip_adapter.py [hunk 3]
```diff
@@ -192,24 +198,46 @@ def _init_image_proj_model(self, state_dict):
...
+def load_ip_adapter_tensors(ip_adapter_ckpt_path: str, device: str) -> IPAdapterStateDict:
+    state_dict: IPAdapterStateDict = {"ip_adapter": {}, "image_proj": {}}
+
+    if ip_adapter_ckpt_path.endswith("safetensors"):
+        model = safe_open(ip_adapter_ckpt_path, device=device, framework="pt")
+        for key in model.keys():
+            if key.startswith("image_proj."):
+                state_dict["image_proj"][key.replace("image_proj.", "")] = model.get_tensor(key)
+            elif key.startswith("ip_adapter."):
+                state_dict["ip_adapter"][key.replace("ip_adapter.", "")] = model.get_tensor(key)
```
Comment (RyanJDick) [inline invokeai/backend/ip_adapter/ip_adapter.py:210]: Add an `else` condition so that we fail loudly if we try to load a model that does not match our assumptions.
```suggestion
            if key.startswith("image_proj."):
                state_dict["image_proj"][key.replace("image_proj.", "")] = model.get_tensor(key)
            elif key.startswith("ip_adapter."):
                state_dict["ip_adapter"][key.replace("ip_adapter.", "")] = model.get_tensor(key)
            else:
                raise RuntimeError(f"Encountered unexpected IP Adapter state dict key: '{key}'.")
```

File: invokeai/backend/ip_adapter/ip_adapter.py [hunk 4]
```diff
@@ -192,24 +198,46 @@ def _init_image_proj_model(self, state_dict):
         ).to(self.device, dtype=self.dtype)
 
 
+def load_ip_adapter_tensors(ip_adapter_ckpt_path: str, device: str) -> IPAdapterStateDict:
+    state_dict: IPAdapterStateDict = {"ip_adapter": {}, "image_proj": {}}
+
+    if ip_adapter_ckpt_path.endswith("safetensors"):
+        model = safe_open(ip_adapter_ckpt_path, device=device, framework="pt")
+        for key in model.keys():
```
Comment (RyanJDick) [inline invokeai/backend/ip_adapter/ip_adapter.py:206]: I suspect that there are also IP-Adapter models saved in safetensors format with the same key structure as `ip_adapter.bin`. Fine with me if you want to wait until we encounter one in the wild though 🙂

**Review by RyanJDick — COMMENTED** [review]
Selecting the wrong CLIP vision encoder for an IP-Adapter causes inference to fail with a dimension mismatch error. We could probably give a clearer error in this case. Or, taking it a step further, we might be able to infer which of ViT-H or ViT-G is compatible with the IP-Adapter before we run it. Totally up to you if you want to tackle this now, or leave it for future work.

**Review by lstein — APPROVED** [review]
Looks OK to me.

File: invokeai/backend/ip_adapter/ip_adapter.py (hunk 3 above)
Comment (blessedcoolant) [reply invokeai/backend/ip_adapter/ip_adapter.py:210]: Do we need to? Because as long as the IP Adapter model has image projection tensors and adapter tensors, we can load it in as a valid model. Any extra keys that might exist in a future model would probably require a custom projection model anyway for the extra functionality but the core aspect of the IP Adapter should still be valid otherwise -- albeit it won't work like advertised but should still boot in.

I see why you want it added though. It would allow us to tightly control the state_dict for any fail case scenario on a custom model but I feel leaving this open here might be better for models that have lingering keys that are not deal breakers for the model itself.

File: invokeai/backend/ip_adapter/ip_adapter.py (hunk 4 above)
Comment (blessedcoolant) [reply invokeai/backend/ip_adapter/ip_adapter.py:206]: What is causing the change in the dict structure? The saving to safetensors? In which case, any bin model converted to safetensors using the native lib should always have the same structure -- unless some madman decides to mess with it. One battle at a time.

File: invokeai/app/invocations/ip_adapter.py (hunk 2 above)
Comment (blessedcoolant) [reply invokeai/app/invocations/ip_adapter.py:69]: The issue here is that the setting is currently mandatory for checkpoint files but not required for the InvokeAI format. Maybe I make it so that the setting is used if the user explicitly sets it - no matter the format but if unset then the former takes over?

File: invokeai/app/invocations/ip_adapter.py [hunk 5]
```diff
@@ -102,19 +114,25 @@ def invoke(self, context: InvocationContext) -> IPAdapterOutput:
...
-                )
-                context.logger.warning("Downloading and installing now. This may take a while.")
-                installer = context._services.model_manager.install
-                job = installer.heuristic_import(f"InvokeAI/{image_encoder_model_name}")
-                installer.wait_for_job(job, timeout=600)  # wait up to 10 minutes - then raise a TimeoutException
-        assert len(image_encoder_models) == 1
+
+            if len(image_encoder_models) == 0:
+                context.logger.error("Error while fetching CLIP Vision Image Encoder")
+                assert len(image_encoder_models) == 1
```
[the remark this answers is not in the document]
Comment (blessedcoolant) [reply invokeai/app/invocations/ip_adapter.py:136]: Ye not a big of how I did this but essentially we're saying "after the model is downloaded, if by any chance the image encoder list is still 0, then we just raise the assertion and end the function" .. because the output of the function is set to get the first item in the retrieved list which will not exist if the download fails. Probably can be done cleaner.

Also not a fan of the fact that I need to call on the `search_by_attrs` twice .. but I need to check if the `ModelInstallJob` has some kind of callback on it. Did this in a rush, so didn't find the time to check if there was one.

File: invokeai/frontend/web/src/services/api/schema.ts [hunk 6]
```diff
@@ -1248,6 +1248,39 @@ export type components = {
        */
       type: "boolean_output";
     };
+    /**
+     * BRIA AI Background Removal
+     * @description Uses the new Bria 1.4 model to remove backgrounds from images.
+     */
+    BriaRemoveBackgroundInvocation: {
```
[the remark this answers is not in the document]
Comment (blessedcoolant) [reply invokeai/frontend/web/src/services/api/schema.ts:1255]: Ah. Nice catch. I'll fix it up.

File: invokeai/backend/model_manager/config.py [hunk 7]
```diff
@@ -323,10 +323,13 @@ def get_tag() -> Tag:
...
 
 
-class IPAdapterConfig(ModelConfigBase):
-    """Model config for IP Adaptor format models."""
-
+class IPAdapterBaseConfig(ModelConfigBase):
     type: Literal[ModelType.IPAdapter] = ModelType.IPAdapter
+
+
+class IPAdapterDiffusersConfig(IPAdapterBaseConfig):
```
[the remark this answers is not in the document]
Comment (blessedcoolant) [reply invokeai/backend/model_manager/config.py:330]: Will change it.

[the remark this answers is not in the document]
Comment (blessedcoolant) [reply invokeai/backend/model_manager/config.py:330]: > taking it a step further, we might be able to infer which of ViT-H or ViT-G is compatible with the IP-Adapter before we run it.

This is the ideal solution but I could not find a way to determine this. Did you give that a shot anytime?

[the remark this answers is not in the document]
Comment (blessedcoolant) [reply invokeai/backend/model_manager/config.py:330]: > Selecting the wrong CLIP vision encoder for an IP-Adapter causes inference to fail with a dimension mismatch error. 

I'll raise a better error message here.

File: invokeai/backend/ip_adapter/ip_adapter.py (hunk 4 above)
Comment (RyanJDick) [reply invokeai/backend/ip_adapter/ip_adapter.py:206]: I'm pretty sure the change in dict structure and saving to safetensors are totally unrelated. But yes, we can wait until this comes up

File: invokeai/backend/ip_adapter/ip_adapter.py (hunk 3 above)
Comment (RyanJDick) [reply invokeai/backend/ip_adapter/ip_adapter.py:210]: I feel like it's better to fail loudly if we receive unexpected keys. It means that either there is a bug in our handling code, or that there is a new feature that we need to add support for. I don't see any advantage to quietly running with an incomplete model - this just seems like it's setting us up for hard-to-debug "Why does it look bad?" issues.

File: invokeai/app/invocations/ip_adapter.py (hunk 2 above)
Comment (RyanJDick) [reply invokeai/app/invocations/ip_adapter.py:69]: I think it's fine as-implemented - just needs a little more explanation in the `description` field.

I suspect that it is possible to figure out the CLIP model, given that choosing the wrong one fails. But, it might be hard. Let's leave that as a future improvement.

**Comment by blessedcoolant** [comment]
Made more or less all the changes we discussed. @RyanJDick Requesting re-review. Gracias.
